### DISCUSSION TITLE
Neovim: build and distribute parsers

### DISCUSSION THREAD

**Discussion by clason** [Homebrew/discussions#3611] [Writing Formulae/Casks]
It seems like Homebrew previously did not build and distribute the C parser as part of the installation. However, since https://github.com/neovim/neovim/pull/15391, tree-sitter-parsers (C, Lua, Viml) are now a mandatory part of the Neovim distribution, which means that `--HEAD` (and 0.8 when that releases in a few weeks) has to be considered broken.

The default build installs these in `/usr/local/lib/nvim/parsers`, although they will be found anywhere in `runtimepath` (including `/usr/local/share/nvim/runtime/parsers`). But since ordering matters, it would be unfortunate if homebrew diverged from the canonical location.

Note that we will possibly add more parsers during the 0.9 release cycle, so it would be better not to hard-code specific parsers. Also, while we endeavor to stick to tagged versions for releases, we may use pre-release commits for nightly and ship corresponding queries in runtime (which may be incompatible with releases).

@carlocab

**Comment by carlocab** [comment]
I did notice that we don't bundle the C parser, which seems to be the only bundled parser in the latest nightly build at the time of writing.

No one seems to have missed them up until now. I have the C, Lua, and Vim parsers installed via nvim-treesitter so I guess I wouldn't have noticed needing to bundle them either.

I'm not opposed to bundling them or shipping them as separate formulae and adding them as `head`-only dependencies, though. The latter might make it easier to do `brew install --HEAD nvim-treesitter-c` if one finds the need for an unreleased parsers. I'd have to think about how to hook separately-installed parsers into `rtp`, though.

**Reply by carlocab** [reply]
We could maybe ship a site configuration file that adds `$HOMEBREW_PREFIX/lib/parsers` to `rtp`.

**Reply by clason** [reply]
I would prefer it if Homebrew considered these as part of the bundled Neovim runtime and include it in the build step, for the above mentioned reasons.

In particular, 
> brew install --HEAD nvim-treesitter-c if one finds the need for an unreleased parsers. 

is not necessarily safe, since queries may be compatible only with a specific commit. (This situation will be rare but possible, and apparently `neovim --HEAD` has quite a lot of users.)

Alternative is to lock and update the parser formulas manually when Neovim HEAD does so, but that seems unsustainable to me.

**Reply by carlocab** [reply]
The difficulty with bundling them with Neovim is that one of three things is likely to happen:
1. They get locked to the versions that work for the released version of Neovim.
2. The released version has parsers locked to what works for it, and `--HEAD` will just pull from Git `HEAD` of parser repos.
3. We ship the parsers only for `--HEAD`, and, depending on how often they get updated, other maintainers start asking questions about why there's so much code churn for `--HEAD`, and maybe we should consider removing it. (Or they also pull from Git `HEAD`.)

> This situation will be rare but possible, and apparently `neovim --HEAD` has quite a lot of users.

It does -- about 70k installs in the past year. That _may_ justify a small maintenance burden, and I am just overly paranoid about item 3 above.

**Reply by clason** [reply]
> The difficulty with bundling them with Neovim is that one of three things is likely to happen:

I'm sorry, but that all seems to be things that would _not_ happen if you bundled the parsers with Neovim (i.e., just let the build script download and build the right version), only if you packaged and distributed them separately?

(And exactly why I was arguing against packaging them individually ;))

**Reply by carlocab** [reply]
Sorry, this is what I meant by bundling: https://github.com/Homebrew/homebrew-core/blob/a8a8f4eb7fe136217988704ba0aa325bc4a2407f/Formula/neovim.rb#L53-L71

We don't usually let build scripts download things on its own and not declared in the formula. (And by "not usually", I mean existing formulae that have build scripts that do that are just oversights.)

**Reply by clason** [reply]
I see. That will be a problem then. In that case, you'll probably have to separate them into HEAD and stable blocks and live with the churn (or disable the `HEAD` build, or put it under a fat disclaimer that it comes without parsers -- a situation I would find preferable to broken builds).

**Reply by carlocab** [reply]
> put it under a fat disclaimer that it comes without parsers

This seems like the simplest thing to do for now... if users complain, then we can entertain other ideas. Feel free to point said complaints to this thread so they can add to the discussion.

**Reply by clason** [reply]
OK, that gives us a few weeks to figure out how to deal with 0.8 (expected September 30), where "Lua treesitter highlighting out of the box" is a marquee feature...

**Reply by carlocab** [reply]
Including parsers for the stable release is easy -- they can be `resource`s or other formulae. The disclaimer is probably going to be there only for `--HEAD`. So you can still make good on the promise of your marquee feature even for those misguided enough to `brew install neovim` 🙂

**Reply by clason** [reply]
That sounds like a reasonable compromise. If you want to track the development branch, you should really build from source yourself anyway (or use the provided nightlies) to help with issues.

**Comment by clason** [comment]
> I did notice that we don't bundle the C parser, which seems to be the only bundled parser in the latest nightly build at the time of writing.

> No one seems to have missed them up until now. I have the C, Lua, and Vim parsers installed via nvim-treesitter so I guess I wouldn't have noticed needing to bundle them either.

Yes, they were not used by Neovim core. That has changed now. (The goal is to support treesitter highlighting _without_ requiring an external plugin.)

(I'm not complaining about the old status quo; just interested in getting the homebrew build up to current spec.)

**Comment by carlocab** [comment]
I've opened Homebrew/homebrew-core#109966 to throw a warning about the not-included parsers for `--HEAD` installs.

**Comment by carlocab** [comment]
How do other distros handle the distribution of parsers? If they're all doing similar things, it may make sense to just follow them, so thinking about downstream isn't too complicated for Neovim.

**Reply by clason** [reply]
I'm not aware of any distribution packaging parsers individually. I think most editors that use tree-sitter (like Helix) statically link them during build?

**Reply by carlocab** [reply]
Hmmm, ok. How does Neovim handle multiple parsers in `rtp`? E.g. what if I have two `parser/c.so`s in my `rtp`? Is the first used? The last? Or something else?

**Reply by clason** [reply]
The first (same as for queries), which typically implies the ordering "user config > plugins > bundled runtime".

**Reply by diktomat** [reply]
As Helix was mentioned, their way for building from source / using head is to have a separate cmd flag for building treesitter: `hx --grammar fetch && hx --grammar build`. While probably no good for releases, Neovim adding a similar command could maybe solve brew --HEAD installation?

**Reply by clason** [reply]
That model doesn't quite apply to Neovim:
* the Lua, Viml, and C parsers are considered a mandatory part of the runtime -- a build not including them is incomplete and not supported
* other parsers are not handled by Neovim at all and must be installed either manually or via a plugin.

(Treesitter support is still unstable, and the question of general parser distribution has not been tackled yet -- we're waiting what upstream will do in this regard first (probably WASM parsers...))

**Reply by baodrate** [reply]
@carlocab neovim now automatically enables treesitter-based syntax highlighting for its help files, by default. (as of neovim/neovim#26347)

this results in an ugly error, on a fresh install without any user configs (using `--headless` here to copy the error but of course happens in the UI):
```
> nvim --clean --headless +:h
Error detected while processing command line..BufReadPost Autocommands for "*":
Error executing lua callback: ...llar/neovim/HEAD-52b6a9a/share/nvim/runtime/filetype.lua:30: Error executing lua: ...llar/neovim/HEAD-52b6a9a/share/nvim/runtime/filetype.lua:31: command line..BufReadPost Autocommands for "*"..FileType Autocommands for "*"..function <SNR>1_LoadFTPlugin[20]..script /opt/homebrew/Cellar/neovim/HEAD-52b6a9a/share/nvim/runtime/ftplugin/help.lua: Vim(runtime):E5113: Error while calling lua chunk: ...b6a9a/share/nvim/runtime/lua/vim/treesitter/language.lua:102: no parser for 'vimdoc' language, see :help treesitter-parsers
stack traceback:
        [C]: in function 'error'
        ...b6a9a/share/nvim/runtime/lua/vim/treesitter/language.lua:102: in function 'add'
        ...a/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:113: in function 'new'
        ...m/HEAD-52b6a9a/share/nvim/runtime/lua/vim/treesitter.lua:41: in function '_create_parser'
        ...m/HEAD-52b6a9a/share/nvim/runtime/lua/vim/treesitter.lua:108: in function 'get_parser'
        ...m/HEAD-52b6a9a/share/nvim/runtime/lua/vim/treesitter.lua:445: in function 'start'
        ...neovim/HEAD-52b6a9a/share/nvim/runtime/ftplugin/help.lua:2: in main chunk
        [C]: in function 'nvim_cmd'
        ...llar/neovim/HEAD-52b6a9a/share/nvim/runtime/filetype.lua:31: in function <...llar/neovim/HEAD-52b6a9a/share/nvim/runtime/filetype.lua:30>
        [C]: in function 'nvim_buf_call'
        ...llar/neovim/HEAD-52b6a9a/share/nvim/runtime/filetype.lua:30: in function <...llar/neovim/HEAD-52b6a9a/share/nvim/runtime/filetype.lua:10>
stack traceback:
        [C]: in function 'nvim_cmd'
        ...llar/neovim/HEAD-52b6a9a/share/nvim/runtime/filetype.lua:31: in function <...llar/neovim/HEAD-52b6a9a/share/nvim/runtime/filetype.lua:30>
        [C]: in function 'nvim_buf_call'
        ...llar/neovim/HEAD-52b6a9a/share/nvim/runtime/filetype.lua:30: in function <...llar/neovim/HEAD-52b6a9a/share/nvim/runtime/filetype.lua:10>
stack traceback:
        [C]: in function 'nvim_buf_call'
        ...llar/neovim/HEAD-52b6a9a/share/nvim/runtime/filetype.lua:30: in function <...llar/neovim/HEAD-52b6a9a/share/nvim/runtime/filetype.lua:10>
```

this makes `--HEAD` installs of neovim annoying to use because there's not an easy way to disable treesitter here (other than patching `vim.treesitter.start`, as @clason [suggested](https://github.com/neovim/neovim/pull/26347#issuecomment-1912060854))

can this be revisted? You mentioned that there's a policy against letting scripts download artifacts untracked by homebrew. could this be relaxed, if only for the `--HEAD` builds? perhaps:

```diff
diff --git a/Formula/n/neovim.rb b/Formula/n/neovim.rb
index 413783e42e0..c1c34015a65 100644
--- a/Formula/n/neovim.rb
+++ b/Formula/n/neovim.rb
@@ -124,6 +124,12 @@ class Neovim < Formula
           end
         end
       end
+    else
+      system "cmake", "-S", "cmake.deps", "-B", ".deps",
+                      "-DUSE_BUNDLED=false",
+                      "-DUSE_BUNDLED_TS_PARSERS=true",
+                      *std_cmake_args
+      system "cmake", "--build", ".deps"
     end
 
     # Point system locations inside `HOMEBREW_PREFIX`.
@@ -148,15 +154,6 @@ class Neovim < Formula
     system "cmake", "--install", "build"
   end
 
-  def caveats
-    return if latest_head_version.blank?
-
-    <<~EOS
-      HEAD installs of Neovim do not include any tree-sitter parsers.
-      You can use the `nvim-treesitter` plugin to install them.
-    EOS
-  end
-
   test do
     refute_match "dirty", shell_output("#{bin}/nvim --version")
     (testpath/"test.txt").write("Hello World from Vim!!")
```

**Reply by clason** [reply]
`--HEAD` builds are unsupported, both by Neovim and by Homebrew. If you _have_ to track `master` (because you want to help with development), build from source yourself.

**Reply by baodrate** [reply]
that's exactly what I'm doing. I'm just using the formula to build from source so homebrew can keep track of the installation. I already committed the change I suggested to my local install. The change would only affect users building w/ `--HEAD`, and since `--HEAD` builds are unsupported anyways I don't see the harm of relaxing the policy against letting the builds download dependencies.

unrelated, but surprisingly a lot of distros' official packages are missing the treesitter parsers. Arch, Alpine, OpenSUSE, Gentoo all package neovim 0.9.5 without them. The packages I've seen that include them so far are homebrew and debian.

**Reply by clason** [reply]
> I'm just using the formula to build from source

Then you are not in fact doing what I'm telling you to do. Clone the repo, `make install`. That will allow you to bisect and apply patches in case of issues.

> The packages I've seen that include them so far are homebrew and debian.

Not a coincidence.

**Reply by baodrate** [reply]
> Not a coincidence.

Not sure what the implication is supposed to be here, sorry.

> Then you are not in fact doing what I'm telling you to do

I think you misunderstand, I'm not asking for support. I've already solved my issue. I just think that if the formula contains a HEAD it should be built properly, but it's not a demand or a request. It's a suggestion. And I'm only offering it because that's specifically what @carlocab [asked for](https://github.com/orgs/Homebrew/discussions/3611#discussioncomment-3568666):

> if users complain, then we can entertain other ideas. Feel free to point said complaints to this thread so they can add to the discussion

If that is unwanted the maintainers can feel free to ignore it. it's no sweat off my back

**Comment by timhillgit** [comment]
As a point of rare agreement between vim and emacs users, it would also be nice for homebrew's emacs formula and cask if the tree-sitter parsers were packaged for homebrew. Emacs _does_ have a command to install the libraries but it requires several manual steps or at the very least some copy-pasting from the internet and trusting it works.

Would the homebrew team be open to the idea of including the parsers with the `tree-sitter` formula? It would look something like adding a resource block for each grammar:
```ruby
  resource "tree-sitter-c" do
    url "https://github.com/tree-sitter/tree-sitter-c.git", tag: "v0.20.8"
       revision: "b125bec9b331722e4eaab6e02e1fe5b14eefd967"
  end
```
followed by a section in `install`:
```ruby
    resources.each do |r|
      libname = "lib#{r.name}.dylib"
      r.stage do
        cd "src" do
          system ENV.cc, "-fPIC", "-c", "-I.", "parser.c"
          system ENV.cc, "-fPIC", "-c", "-I.", "scanner.c"
          system ENV.cc, "-fPIC", "-shared", "*.o", "-o", libname
          lib.install libname
        end
      end
    end
```

**Reply by clason** [reply]
However much I like to agree on _something_, that's the opposite of what I'm asking for... (Bundled queries are only guaranteed to be compatible with specific parser revisions, so we can't trust random libs on the filesystem.)

**Reply by timhillgit** [reply]
Alas! Sorry about the misread. Potentially emacs could follow your lead and bundle the libraries in the formula and cask though that does seem like a lot of duplication. At this point I think it comes down to what the maintainers will accept?

**Reply by clason** [reply]
You could make their life easier by just using Neovim's queries (to guarantee you can use the same parser version) ;)
